src/Operations.jl

```diff
@@ -1582,45 +1582,56 @@ function diff_array(old_ctx::Union{Context,Nothing}, new_ctx::Context; manifest=
 end
 
+is_package_downloaded(ctx, pkg::PackageSpec) = isdir(project_rel_path(ctx, source_path(pkg)))
+
 function print_status(ctx::Context, old_ctx::Union{Nothing,Context}, header::Symbol,
                       uuids::Vector, names::Vector; manifest=true, diff=false)
     ctx.io = something(ctx.status_io, ctx.io) # for instrumenting tests
     filter = !isempty(uuids) || !isempty(names)
-    # print header
-    printpkgstyle(ctx, header, pathrepr(manifest ? ctx.env.manifest_file : ctx.env.project_file))
     # setup
-    printed_something, empty_diff = false, true
     xs = diff_array(old_ctx, ctx; manifest=manifest)
-    xs = sort!(xs, by = (x -> (is_stdlib(x[1]), something(x[3], x[2]).name, x[1])))
-    # return early
+    # filter and return early if possible
+    if isempty(xs) && !diff
+        printpkgstyle(ctx, header, "$(pathrepr(manifest ? ctx.env.manifest_file : ctx.env.project_file)) (empty " *
+                      (manifest ? "manifest" : "project") * ")")
+        return nothing
+    end
+    xs = !diff ? xs : [(id, old, new) for (id, old, new) in xs if old != new]
     if isempty(xs)
-        printpkgstyle(ctx, header, manifest ? "empty manifest" : "empty project")
+        printpkgstyle(ctx, Symbol("No Changes"), "to $(pathrepr(manifest ? ctx.env.manifest_file : ctx.env.project_file))")
+        return nothing
+    end
+    xs = !filter ? xs : [(id, old, new) for (id, old, new) in xs if (id in uuids || something(new, old).name in names)]
+    if isempty(xs)
+        printpkgstyle(ctx, Symbol("No Matches"),
+                      "in $(diff ? "diff for " : "")$(pathrepr(manifest ? ctx.env.manifest_file : ctx.env.project_file))")
         return nothing
     end
     # main print
+    printpkgstyle(ctx, header, pathrepr(manifest ? ctx.env.manifest_file : ctx.env.project_file))
+    xs = sort!(xs, by = (x -> (is_stdlib(x[1]), something(x[3], x[2]).name, x[1])))
+    all_packages_downloaded = true
     for (uuid, old, new) in xs
-        diff && old == new && continue # in diff mode and no diff to show
-        empty_diff = false
-        filter && !(uuid in uuids) && !(something(new, old).name in names) && continue
-        printed_something = true
-
-        printstyled(ctx.io, "   $(string(uuid)[1:8]) "; color = :light_black)
+        pkg_downloaded = !is_instantiated(new) || is_package_downloaded(ctx, new)
+        all_packages_downloaded &= pkg_downloaded
+        printstyled(ctx.io, pkg_downloaded ? " " : "→"; color = :red)
+        printstyled(ctx.io, " [", string(uuid)[1:8], "] "; color = :light_black)
         diff ? print_diff(ctx, old, new) : print_single(ctx, new)
         println(ctx.io)
     end
-    if !printed_something
-        printpkgstyle(ctx, header, (diff && empty_diff) ? "empty diff" : "no matches")
+    if !all_packages_downloaded
+        printpkgstyle(ctx, :Info, "packages marked with → are not downloaded, use `instantiate` to download", true)
```

Comment:
I think there is some extra clarity to do it the way I did in https://github.com/JuliaLang/Pkg.jl/pull/1656/files#diff-920ed338dcbd9d62489c6d12dc2fbb18R1636 to show the arrow as red.

Reply:
ah, yes that is definitely better